File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> prefixSumCount;
        prefixSumCount[0] = 1; // to handle subarrays starting at index 0

        int currSum = 0;
        int count = 0;

        for (int num : nums) {
            currSum += num;

            // check if (currSum - k) exists
            if (prefixSumCount.find(currSum - k) != prefixSumCount.end()) {
                count += prefixSumCount[currSum - k];
            }

            // store current prefix sum
            prefixSumCount[currSum]++;
        }

        return count;
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer,Integer> mp=new HashMap<>();
        mp.put(0,1);
        int n=nums.length;
        int count=0;
        int curr=0;
        for(int i=0;i<n;i++){
            curr+=nums[i];

            if(mp.containsKey(curr-k)){
                count+=mp.get(curr-k);
            }
            mp.put(curr,mp.getOrDefault(curr,0)+1);
        }
        return count;
    }
}